app/uk/gov/hmrc/transitmovementsrouter/controllers/stream/StreamingParsers.scala

```diff
@@ -52,16 +56,35 @@ trait StreamingParsers extends StreamWithFile with FutureConversions {
 
   implicit class ActionBuilderStreamHelpers(actionBuilder: ActionBuilder[Request, _]) {
 
-    def stream(
+    // This method allows for the transformation of a stream before it goes into a file,
+    // such that we only transform it once.
+    def stream(transformer: Flow[ByteString, ByteString, _])(
       block: Request[Source[ByteString, _]] => Future[Result]
     )(implicit temporaryFileCreator: TemporaryFileCreator): Action[Source[ByteString, _]] =
       actionBuilder.async(streamFromMemory) {
         request =>
-          withReusableSource(request.body) {
-            memoryOrFileSource =>
-              block(request.withBody(memoryOrFileSource))
-          }
+          val tempFile = temporaryFileCreator.create()
+          request.body
+            .via(transformer)
+            .runWith(FileIO.toPath(tempFile))
+            .transformWith {
+              case Success(_) => block(request.withBody(FileIO.fromPath(tempFile)))
+              case Failure(error: IOOperationIncompleteException)
+                  if error.getCause.isInstanceOf[IllegalStateException] || error.getCause.isInstanceOf[WFCException] =>
+                Future.successful(Status(BAD_REQUEST)(Json.toJson(PresentationError.badRequestError(error.getCause.getMessage))))
```

Comment:
Would it be a bad request if the transformation done by us causes the failure/how do we end up getting this failure?

Reply:
The bad request is for catching incorrect incoming XML -- those errors are caused when parsing the initial XML, so bad request is correct here.